File: Languages/go/algorithms-dsa/leetcode/heap/find_median_data_stream.go
package main

import (
	"container/heap"
	"fmt"
)

/*
  LC 295 - Find Median from Data Stream
  Topic: Heap (Two Heaps)
  Difficulty: Hard
  Time: O(log n) add, O(1) find | Space: O(n)
*/

type MaxHeap []int

func (h MaxHeap) Len() int            { return len(h) }
func (h MaxHeap) Less(i, j int) bool   { return h[i] > h[j] }
func (h MaxHeap) Swap(i, j int)        { h[i], h[j] = h[j], h[i] }
func (h *MaxHeap) Push(x interface{}) { *h = append(*h, x.(int)) }
func (h *MaxHeap) Pop() interface{} {
	old := *h
	x := old[len(old)-1]
	*h = old[:len(old)-1]
	return x
}

type MinHeap []int

func (h MinHeap) Len() int            { return len(h) }
func (h MinHeap) Less(i, j int) bool   { return h[i] < h[j] }
func (h MinHeap) Swap(i, j int)        { h[i], h[j] = h[j], h[i] }
func (h *MinHeap) Push(x interface{}) { *h = append(*h, x.(int)) }
func (h *MinHeap) Pop() interface{} {
	old := *h
	x := old[len(old)-1]
	*h = old[:len(old)-1]
	return x
}

type MedianFinder struct {
	lo *MaxHeap // max-heap for lower half
	hi *MinHeap // min-heap for upper half
}

func Constructor() MedianFinder {
	lo, hi := &MaxHeap{}, &MinHeap{}
	return MedianFinder{lo, hi}
}

func (mf *MedianFinder) AddNum(num int) {
	heap.Push(mf.lo, num)
	heap.Push(mf.hi, heap.Pop(mf.lo))
	if mf.hi.Len() > mf.lo.Len() {
		heap.Push(mf.lo, heap.Pop(mf.hi))
	}
}

func (mf *MedianFinder) FindMedian() float64 {
	if mf.lo.Len() > mf.hi.Len() {
		return float64((*mf.lo)[0])
	}
	return float64((*mf.lo)[0]+(*mf.hi)[0]) / 2.0
}

func main() {
	mf := Constructor()
	mf.AddNum(1)
	mf.AddNum(2)
	if mf.FindMedian() == 1.5 {
		fmt.Println("PASS: median after [1,2]")
	} else {
		fmt.Println("FAIL: median after [1,2]")
	}
	mf.AddNum(3)
	if mf.FindMedian() == 2.0 {
		fmt.Println("PASS: median after [1,2,3]")
	} else {
		fmt.Println("FAIL: median after [1,2,3]")
	}
}

File: Languages/cpp/algorithms-dsa/leetcode/heap/find_median_data_stream.cpp
/*
 * LeetCode 295 - Find Median from Data Stream
 * Topic: Heap
 * Difficulty: Hard
 *
 * Two heaps: max-heap for lower half, min-heap for upper half. Balance sizes.
 * Time: O(log n) per addNum, O(1) per findMedian
 * Space: O(n)
 */
#include <iostream>
#include <queue>
#include <cassert>
#include <cmath>
using namespace std;

class MedianFinder {
    priority_queue<int> lo;                              // max-heap
    priority_queue<int, vector<int>, greater<int>> hi;   // min-heap
public:
    void addNum(int num) {
        lo.push(num);
        hi.push(lo.top()); lo.pop();
        if (hi.size() > lo.size()) { lo.push(hi.top()); hi.pop(); }
    }
    double findMedian() {
        return lo.size() > hi.size() ? lo.top() : (lo.top() + hi.top()) / 2.0;
    }
};

int main() {
    MedianFinder mf;
    mf.addNum(1);
    assert(fabs(mf.findMedian() - 1.0) < 1e-5);
    mf.addNum(2);
    assert(fabs(mf.findMedian() - 1.5) < 1e-5);
    mf.addNum(3);
    assert(fabs(mf.findMedian() - 2.0) < 1e-5);
    mf.addNum(4);
    assert(fabs(mf.findMedian() - 2.5) < 1e-5);

    cout << "All tests passed!" << endl;
    return 0;
}
